### PR TITLE
scx_layered: fix stalls w/ affinitized threads

### PR DESCRIPTION
**Pull request by likewhatevs**
latest commit is the one that ran for 5 mins w/ `stress-ng -c 128 -M --affinity 128` and no stall

Alacritty 2024-10-31 10 06 51

don't understand it yet but think there's something to this commit (because 5 minutes and no stall, commit before it stalled).

### CODE REVIEW

**Comment by likewhatevs** [comment]
gonna play with this as idk baseline to see what change exactly did it

**Comment by likewhatevs** [comment]
did a cargo clean and rebuild and yeah somehow that doesn't stall:

CleanShot 2024-10-31 at 10 21 34@2x

**Comment by likewhatevs** [comment]
latest commit (clean, build, `stress-ng -c 128 -M --affinity 128`) works still:

Alacritty 2024-10-31 10 41 45

config is:
```
[
  {
    "name": "random",
    "comment": "random threads",
    "matches": [
      [
        {
          "CommPrefix": "dontmatter"
        }
      ],
      [
        {
          "CommPrefix": "DONTMATTER"
        }
      ]
    ],
    "kind": {
      "Open": {
        "min_exec_us": 0,
        "slice_us": 1000,
        "preempt": false,
        "exclusive": false,
        "growth_algo": "Random",
        "weight": 200
      }
    }
},
  {
    "name": "all",
    "comment": "all threads",
    "matches": [
      []
    ],
    "kind": {
      "Open": {
        "min_exec_us": 50,
        "slice_us": 1000,
        "preempt": false,
        "exclusive": false,
        "growth_algo": "Random",
        "weight": 100
      }
    }
  }
]
```

**Comment by likewhatevs** [comment]
removing this change brought back stalls:

Alacritty 2024-10-31 10 57 39

```
  R stress-ng-affin[113267] -38020ms
      scx_state/flags=3/0x1 dsq_flags=0x1 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=0x3 dsq_vtime=525847781
      cpus=00000001

    __set_cpus_allowed_ptr+0x71/0xd0
    __sched_setaffinity+0xdf/0x1b0
    sched_setaffinity+0x1cc/0x290
    __x64_sys_sched_setaffinity+0xa6/0xc0
    do_syscall_64+0x82/0x180
    entry_SYSCALL_64_after_hwframe+0x4b/0x53

  R stress-ng-affin[113279] -38020ms
      scx_state/flags=3/0x1 dsq_flags=0x1 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=0x3 dsq_vtime=525847781
      cpus=00000001

    __set_cpus_allowed_ptr+0x71/0xd0
    __sched_setaffinity+0xdf/0x1b0
    sched_setaffinity+0x1cc/0x290
    __x64_sys_sched_setaffinity+0xa6/0xc0
    do_syscall_64+0x82/0x180
    entry_SYSCALL_64_after_hwframe+0x4b/0x53

  R stress-ng-affin[113290] -38020ms
      scx_state

~~~~ TRUNCATED ~~~~

================================================================================

Error: EXIT: runnable task stall (stress-ng-affin[112972] failed to run for 38.047s)

```

**Comment by likewhatevs** [comment]
also (misc btw) topology awareness autodetect seems wrong (i.e. 9950x should have topo off right)?

File: scheds/rust/scx_layered/src/bpf/main.bpf.c [hunk 1]
```diff
@@ -1060,6 +1062,10 @@ void BPF_STRUCT_OPS(layered_enqueue, struct task_struct *p, u64 enq_flags)
 	if (vtime_before(vtime, layer->vtime_now - layer_slice_ns))
 		vtime = layer->vtime_now - layer_slice_ns;
 
+	if (!tctx->all_cpus_allowed){
```
Comment (likewhatevs) [inline scheds/rust/scx_layered/src/bpf/main.bpf.c:1065]: topo being wrong means that all that is needed to fix stalls might be this 4 lines, testing w/ topo disabled to see if stall or nostall w/o the additional lines.

**Comment by likewhatevs** [comment]
no stalls w/ timer stuff disabled and topo off:

Alacritty 2024-10-31 11 09 28

Alacritty 2024-10-31 11 09 52

need to confirm main shows stalls w/ topo off to like, confirm that fix actually fixes but yeah looks like only those 4 lines do it.

**Comment by likewhatevs** [comment]
still works w and w/o topo, confirming failures on main next:

CleanShot 2024-10-31 at 11 36 44@2x

**Comment by likewhatevs** [comment]
The test case fails on main with topology enabled (so this change fixes at least that). Need to let a test with topology disabled for longer (i.e. it kinda looks like no stall on main atm).

```
❯ sudo ./target/release/scx_layered --disable-topology=false f:/home/patso/repos/new-layered-config.json
15:42:39 [INFO] CPUs: online/possible=32/32 nr_cores=16
15:42:39 [INFO] Layered Scheduler Attached. Run `scx_layered --monitor` for metrics.

DEBUG DUMP
================================================================================
```

**Comment by likewhatevs** [comment]
yeah that path seems fine:
CleanShot 2024-10-31 at 11 57 32@2x

retrying with a different config to see if I can get main to fail the test on both paths and then see how/if the change alters results.

**Comment by likewhatevs** [comment]
Tried with a different config, same results on main, testing changes next:

```
❯ cat ../layered-config.json
[
{
  "name":"patso",
  "comment":"patso user",
  "matches":[
     [{"UIDEquals":1000}]
  ],
  "kind": {
    "Open":{
      "slice_us": 1000,
      "preempt": false,
      "preempt_first": false,
      "exclusive": false,
      "weight": 100
     }
  }
},{
  "name":"stress-ng",
  "comment":"stress-ng slice",
  "matches":[
     [
             {"CommPrefix":"stress-ng"}
     ],[
             {"PcommPrefix":"stress-ng"}
     ]],
  "kind": {
    "Open":{
      "slice_us": 1000,
      "preempt": false,
      "preempt_first": false,
      "exclusive": false,
      "weight": 100
     }
  }
},
{
  "name":"normal",
  "comment":"the rest",
  "matches":[[]],
  "kind":{
    "Open": {
      "slice_us": 1000,
      "preempt": false,
      "preempt_first": false,
      "exclusive": false,
      "weight": 200
    }
  }
}]
```

**Comment by likewhatevs** [comment]
Main reproduces the same thing with the new config.

With the new config, I consistently get:

```
❯ sudo ./target/release/scx_layered --disable-topology=false f:/home/patso/repos/layered-config.json
16:54:11 [INFO] CPUs: online/possible=32/32 nr_cores=16
16:54:11 [INFO] Layered Scheduler Attached. Run `scx_layered --monitor` for metrics.

DEBUG DUMP
================================================================================

stress-ng[15792] triggered exit kind 1024:
  runtime error (DSQ ID 0x0000000000000000 already had FIFO-enqueued tasks)

Backtrace:
  dispatch_enqueue+0x358/0x3d0
  enqueue_task_scx+0x19d/0x2b0
  enqueue_task+0x41/0x1f0
  wake_up_new_task+0x19a/0x310
  kernel_clone+0x26d/0x450
  __do_sys_clone+0x66/0x90
  do_syscall_64+0x82/0x180
  entry_SYSCALL_64_after_hwframe+0x4b/0x53

01234567|89012345|67890123|45678901|


HI_FALLBACK[1024] nr_queued=0 -0ms
HI_FALLBACK[1025] nr_queued=0 -0ms
LO_FALLBACK nr_queued=0 -0ms

CPU states
----------

CPU 5   : nr_run=1 flags=0x1 cpu_rel=0 ops_qseq=32 pnt_seq=209
          curr=screen[2459] class=ext_sched_class

 *R screen[2459] -4ms
      scx_state/flags=3/0xd dsq_flags=0x0 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=(n/a) dsq_vtime=442006790
      cpus=ffffffff

CPU 12  : nr_run=2 flags=0x1 cpu_rel=0 ops_qseq=28 pnt_seq=1032
          curr=ksoftirqd/12[93] class=ext_sched_class

 *R ksoftirqd/12[93] +0ms
      scx_state/flags=3/0x5 dsq_flags=0x0 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=(n/a) dsq_vtime=810397621
      cpus=00001000

  R kworker/12:2[1539] +0ms
      scx_state/flags=3/0x9 dsq_flags=0x0 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=0x0 dsq_vtime=10790182
      cpus=00001000

    kthread+0xcf/0x100
    ret_from_fork+0x4d/0x60
    ret_from_fork_asm+0x11/0x20

CPU 15  : nr_run=2 flags=0x1 cpu_rel=0 ops_qseq=24 pnt_seq=424
          curr=stress-ng[15793] class=ext_sched_class

 *R stress-ng[15793] +0ms
      scx_state/flags=3/0x5 dsq_flags=0x0 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=(n/a) dsq_vtime=939475049
      cpus=ffffffff

  R ksoftirqd/15[111] +0ms
      scx_state/flags=3/0x9 dsq_flags=0x0 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=0x0 dsq_vtime=810397621
      cpus=00008000

    kthread+0xcf/0x100
    ret_from_fork+0x4d/0x60
    ret_from_fork_asm+0x11/0x20

CPU 19  : nr_run=2 flags=0x5 cpu_rel=0 ops_qseq=20 pnt_seq=144
          curr=stress-ng[15792] class=ext_sched_class
  idle_to_kick   : 00000004

 *R stress-ng[15792] -14ms
      scx_state/flags=3/0x5 dsq_flags=0x0 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=(n/a) dsq_vtime=939475049
      cpus=ffffffff

    scx_dump_state+0x680/0x7c0
    scx_ops_error_irq_workfn+0x3d/0x40
    irq_work_single+0x2f/0x80
    irq_work_run_list+0x2c/0x40
    irq_work_run+0x18/0x30
    __sysvec_irq_work+0x1c/0xc0
    sysvec_irq_work+0x7a/0x90
    asm_sysvec_irq_work+0x1a/0x20
    wake_up_new_task+0x217/0x310
    kernel_clone+0x26d/0x450
    __do_sys_clone+0x66/0x90
    do_syscall_64+0x82/0x180
    entry_SYSCALL_64_after_hwframe+0x4b/0x53

  R kworker/19:1H[1429] -7ms
      scx_state/flags=3/0x1 dsq_flags=0x0 ops_state/qseq=0/0
      sticky/holding_cpu=-1/-1 dsq_id=0x0 dsq_vtime=809397621
      cpus=00080000

    kthread+0xcf/0x100
    ret_from_fork+0x4d/0x60
    ret_from_fork_asm+0x11/0x20

================================================================================

Error: EXIT: runtime error (DSQ ID 0x0000000000000000 already had FIFO-enqueued tasks)
```

but seeing as this fixes the stalls with the other config, I think the next thing to do is fix that issue in a manner that keeps the change that reduced stalls.

File: scheds/rust/scx_layered/src/bpf/main.bpf.c [hunk 2]
```diff
@@ -1060,6 +1060,10 @@ void BPF_STRUCT_OPS(layered_enqueue, struct task_struct *p, u64 enq_flags)
...
+		scx_bpf_dispatch(p, idx, slice_ns, enq_flags);
```
Comment (likewhatevs) [inline scheds/rust/scx_layered/src/bpf/main.bpf.c:1064]: also tracking idx is uninitialized, what I don't get is how this resolves stalls with the first config I tested.

**Comment by likewhatevs** [comment]
The latest changes make more sense and prevent errors with the new config. no stalls with that either, testing old config:

CleanShot 2024-10-31 at 13 43 26@2x

**Comment by likewhatevs** [comment]
passes w/ new also
CleanShot 2024-10-31 at 13 48 36@2x

**Comment by likewhatevs** [comment]
TL;DR of how this PR evolved -- just throw affinitized threads on global DSQ more/less immediately (which I guess kind of makes sense if it doesn't cause other issues).

File: scheds/rust/scx_layered/src/bpf/main.bpf.c [hunk 3]
```diff
@@ -1060,6 +1060,10 @@ void BPF_STRUCT_OPS(layered_enqueue, struct task_struct *p, u64 enq_flags)
 	if (vtime_before(vtime, layer->vtime_now - layer_slice_ns))
 		vtime = layer->vtime_now - layer_slice_ns;
 
+	if (!tctx->all_cpus_allowed){
+		scx_bpf_dispatch(p, SCX_DSQ_GLOBAL, slice_ns, enq_flags);
+		goto preempt;
```
Comment (htejun) [inline scheds/rust/scx_layered/src/bpf/main.bpf.c:1065]: So, the problem with this is that on muti-socket machines, tasks which are affine to a NUMA node would all slip out of their layers and end up in `DSQ_GLOBAL`.

Comment (likewhatevs) [reply scheds/rust/scx_layered/src/bpf/main.bpf.c:1065]: Thanks, will see if stalls go away using SCX_DSQ_LOCAL (I think that avoids that?).

**Comment by likewhatevs** [comment]
The tests still pass so this might work?

CleanShot 2024-10-31 at 14 18 47@2x

Updated to use SCX_DSQ_LOCAL instead of SCX_DSQ_LOCAL to avoid issues with NUMA affinities.

File: scheds/rust/scx_layered/src/bpf/main.bpf.c (hunk 3 above)
Comment (htejun) [reply scheds/rust/scx_layered/src/bpf/main.bpf.c:1065]: It'd be avoid `GLOBAL` but that likewise would send all such tasks to the local CPU (and e.g. it wouldn't be able to preempt other tasks if the local CPU is currently busy).

Comment (likewhatevs) [reply scheds/rust/scx_layered/src/bpf/main.bpf.c:1065]: > wouldn't be able to preempt other tasks if the local CPU is currently busy

I'm not 1000% sold on that being the problem I want to solve (yet). 

I previously thought it was, but my test captured an issue similar looking to the one I want to solve which makes me think "maybe this is the problem instead of that", idk.

**Comment by likewhatevs** [comment]
Figured out why this works -- SCX_DSQ_LOCAL is executed before layers are (so the solve all stalls fix is the timer and walk the DSQs one).

Will open a new PR with that because this one got messy.
